toolchains/node/node_configure.bzl

```diff
@@ -0,0 +1,75 @@
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Defines a repository rule for configuring the node binary.
+"""
+
+def _impl(repository_ctx):
+    if repository_ctx.attr.target_tool and repository_ctx.attr.target_tool_path:
+        fail("Can only set one of target_tool or target_tool_path but both where set.")
+
+    host_os = repository_ctx.os.name.lower()
+    substitutions = {}
+    default_tool_path = repository_ctx.attr.target_tool_path or repository_ctx.which("node") or ""
+    substitutions["%{TOOL_ATTRS}"] = "    target_tool_path = \"%s\"\n" % default_tool_path
+
+    if repository_ctx.attr.target_tool:
+        substitutions["%{TOOL_ATTRS}"] = "    target_tool = \"%s\"\n" % repository_ctx.attr.target_tool
+
+    repository_ctx.template(
+        "BUILD",
+        Label("@build_bazel_rules_nodejs//toolchains/node:BUILD.tpl"),
+        substitutions,
+        False,
+    )
+
+_node_configure = repository_rule(
+    implementation = _impl,
+    attrs = {
+        "target_tool": attr.label(
+            doc = "Target for a downloaded nodejs binary for the target os.",
+            mandatory = False,
+            allow_single_file = True,
+        ),
+        "target_tool_path": attr.string(
+            doc = "Absolute path to a pre-installed nodejs binary for the target os.",
+            mandatory = False,
+        ),
+    },
+)
+
+def node_configure(node_repository_names = None, **kwargs):
+    """Creates an external repository with a node_toolchain target properly configured.
+
+    Args:
+      node_repository_names: list of strings
+      **kwargs: all the args of the _node_configure rules
+    """
+
+    native.register_toolchains(
```

Comment:
Seems like the call to `native.register_toolchains(` would fit better in `node_repositories` itself that is aware of all the OS names at and if a user uses `node_configure` then the toolchains in this list won't be registered twice.

Something like this perhaps in node_repositories():
```
    # This needs to be setup so toolchains can access nodejs for all different versions
    for os_arch_name in OS_ARCH_NAMES:
        os_name = "_".join(os_arch_name)
        node_repository_name = "nodejs_%s" % os_name
        _maybe(
            _nodejs_repo,
            name = node_repository_name,
            package_json = package_json,
            node_version = node_version,
            yarn_version = yarn_version,
            vendored_node = vendored_node,
            vendored_yarn = vendored_yarn,
            node_repositories = node_repositories,
            yarn_repositories = yarn_repositories,
            node_urls = node_urls,
            yarn_urls = yarn_urls,
            preserve_symlinks = preserve_symlinks,
        )
        native.register_toolchains("@build_bazel_rules_nodejs//toolchains/node:node_%s_toolchain" % os_arch_name[0])
        node_toolchain_configure(
            name = "%s_config" % node_repository_name,
            target_tool = "@%s//:node_bin" % node_repository_name,
        )
```

which means the `node_configure` macro could go and it would just be `node_configure = repository_rule(` and also the toolchain name for osx would have to change to:

```
toolchain(
    name = "node_darwin_toolchain",
    target_compatible_with = [
        "@bazel_tools//platforms:osx",
        "@bazel_tools//platforms:x86_64",
    ],
    toolchain = "@nodejs_darwin_amd64_config//:toolchain",
    toolchain_type = ":toolchain_type",
)
```
?